edx_lint/pylint/feature_toggle_check.py

```diff
@@ -0,0 +1,186 @@
+
+        """
+        line_start_indexes = [0]
+        index = 0
+        while True:
+            index = string.find('\n', index)
+            if index < 0:
+                break
+            index += 1
+            line_start_indexes.append(index)
+        return line_start_indexes
+
+    def _line_number_to_line(self, line_number):
+        """
+        Gets the line of text designated by the provided line number.
+
+        Arguments:
+            line_number: The line number of the line we want to find.
+
+        Returns:
+            The line of text designated by the provided line number.
+
+        """
+        start_index = self._line_start_indexes[line_number - 1]
+        if len(self._line_start_indexes) == line_number:
+            line = self._string[start_index:]
+        else:
+            end_index = self._line_start_indexes[line_number]
+            line = self._string[start_index:end_index - 1]
+        return line
+
+
+@check_visitors
+class FeatureToggleChecker(BaseChecker):
+    """
+    Checks that feature toggles are properly annotated.
+    """
+
+    __implements__ = (IAstroidChecker,)
+
+    name = 'feature-toggle-annotation-checker'
+
+    WAFFLE_NOT_ANNOTATED_MESSAGE_ID = 'waffle-missing-toggle-annotation'
+    ILLEGAL_WAFFLE_MESSAGE_ID = 'illegal-waffle-usage'
+
+    _CHECK_CAPITAL_REGEX = re.compile(r'[A-Z]')
+    _WAFFLE_TOGGLE_CLASSES = ['WaffleFlag', 'WaffleSwitch',]
+    _ILLEGAL_WAFFLE_FUNCTIONS = ['flag_is_active', 'switch_is_active',]
+
+    msgs = {
+        'E%d40' % BASE_ID: (
+            u"waffle feature toggle (%s) is missing annotation",
+            WAFFLE_NOT_ANNOTATED_MESSAGE_ID,
+            "waffle feature toggle is missing annotation",
+        ),
+        'E%d41' % BASE_ID: (
+            u"illegal waffle usage with (%s). use utility classes {}.".format(', '.join(_WAFFLE_TOGGLE_CLASSES)),
+            ILLEGAL_WAFFLE_MESSAGE_ID,
+            "illegal waffle usage. use utility classes {}.".format(', '.join(_WAFFLE_TOGGLE_CLASSES)),
+        ),
+    }
+
+    def __init__(self, *args, **kwargs):
+        super(FeatureToggleChecker, self).__init__(*args, **kwargs)
+        self._lines = None
+
+    def visit_module(self, node):
+        """Parses the module code to provide access to comments."""
+        self._lines = AnnotationLines(node)
+
+    def check_waffle_class_annotated(self, node):
+        """
+        Check Call node for waffle class instantiation with missing annotations.
+        """
+        if not hasattr(node.func, 'name'):
+            return
+
+        # Looking for class instantiation, so should start with a capital letter
+        starts_with_capital = self._CHECK_CAPITAL_REGEX.match(node.func.name)
+        if not starts_with_capital:
+            return
+
+        # Search for toggle classes that require an annotation
+        if not any(node.func.name.endswith(toggle_class) for toggle_class in self._WAFFLE_TOGGLE_CLASSES):
```

Comment:
Fun fact: endswith can take a tuple of possibilities, and check them all.

Reply:
👍